### PR TITLE
Ignore gossip messages when in optimistic mode

### PR DESCRIPTION
**Pull request by jmozah**
This PR implements ignoring topic messages when in optimistic mode. Only `blocks` topic is kept open.

### CODE REVIEW

**Review by terencechain — COMMENTED** [review]
Let's target this for `develop` after https://github.com/prysmaticlabs/prysm/pull/10276 is merged

File: beacon-chain/sync/validate_aggregate_proof_test.go [hunk 1]
```diff
@@ -686,3 +686,35 @@ func TestValidateAggregateAndProof_RejectWhenAttEpochDoesntEqualTargetEpoch(t *t
...
+	require.NoError(t, err)
+	topic := p2p.GossipTypeMapping[reflect.TypeOf(exit)]
+	m := &pubsub.Message{
+		Message: &pubsubpb.Message{
+			Data:  buf.Bytes(),
+			Topic: &topic,
+		},
+	}
+	res, err := r.validateAggregateAndProof(ctx, "", m)
+	_ = err
```
Comment (prestonvanloon) [inline beacon-chain/sync/validate_aggregate_proof_test.go:717]:
```suggestion
	assert.NoError(t, err)
```

File: beacon-chain/sync/validate_attester_slashing_test.go [hunk 2]
```diff
@@ -293,3 +293,34 @@ func TestSeenAttesterSlashingIndices(t *testing.T) {
...
+
+	topic := p2p.GossipTypeMapping[reflect.TypeOf(slashing)]
+	msg := &pubsub.Message{
+		Message: &pubsubpb.Message{
+			Data:  buf.Bytes(),
+			Topic: &topic,
+		},
+	}
+	res, err := r.validateAttesterSlashing(ctx, "foobar", msg)
+	_ = err
```
Comment (prestonvanloon) [inline beacon-chain/sync/validate_attester_slashing_test.go:323]:
```suggestion
	assert.NoError(t, err)
```

File: beacon-chain/sync/validate_beacon_attestation_test.go [hunk 3]
```diff
@@ -298,3 +301,34 @@ func TestService_validateCommitteeIndexBeaconAttestation(t *testing.T) {
...
+
+	topic := p2p.GossipTypeMapping[reflect.TypeOf(slashing)]
+	msg := &pubsub.Message{
+		Message: &pubsubpb.Message{
+			Data:  buf.Bytes(),
+			Topic: &topic,
+		},
+	}
+	res, err := r.validateCommitteeIndexBeaconAttestation(ctx, "foobar", msg)
+	_ = err
```
Comment (prestonvanloon) [inline beacon-chain/sync/validate_beacon_attestation_test.go:331]:
```suggestion
	assert.NoError(t, err)
```

File: beacon-chain/sync/validate_proposer_slashing_test.go [hunk 4]
```diff
@@ -209,3 +209,33 @@ func TestValidateProposerSlashing_Syncing(t *testing.T) {
...
+	require.NoError(t, err)
+	topic := p2p.GossipTypeMapping[reflect.TypeOf(slashing)]
+	m := &pubsub.Message{
+		Message: &pubsubpb.Message{
+			Data:  buf.Bytes(),
+			Topic: &topic,
+		},
+	}
+	res, err := r.validateProposerSlashing(ctx, "", m)
+	_ = err
```
Comment (prestonvanloon) [inline beacon-chain/sync/validate_proposer_slashing_test.go:238]:
```suggestion
	assert.NoError(t, err)
```

File: beacon-chain/sync/validate_voluntary_exit_test.go [hunk 5]
```diff
@@ -195,3 +195,35 @@ func TestValidateVoluntaryExit_ValidExit_Syncing(t *testing.T) {
...
+	require.NoError(t, err)
+	topic := p2p.GossipTypeMapping[reflect.TypeOf(exit)]
+	m := &pubsub.Message{
+		Message: &pubsubpb.Message{
+			Data:  buf.Bytes(),
+			Topic: &topic,
+		},
+	}
+	res, err := r.validateVoluntaryExit(ctx, "", m)
+	_ = err
```
Comment (prestonvanloon) [inline beacon-chain/sync/validate_voluntary_exit_test.go:226]:
```suggestion
	assert.NoError(t, err)
```

**Review by prestonvanloon — CHANGES_REQUESTED** [review]
Please handle errors in test

File: beacon-chain/sync/validate_aggregate_proof.go [hunk 6]
```diff
@@ -41,6 +41,16 @@ func (s *Service) validateAggregateAndProof(ctx context.Context, pid peer.ID, ms
 		return pubsub.ValidationIgnore, nil
 	}
 
+	// We should not attempt to process this message if the node is running in optimistic mode.
+	// We just ignore in p2p so that the peer is not penalized.
+	status, err := s.cfg.chain.IsOptimistic(ctx)
+	if err != nil {
+		return pubsub.ValidationReject, nil
```
Comment (nisdas) [inline beacon-chain/sync/validate_aggregate_proof.go:48]: you should return the error here, we have gossip message handler which will log errors for rejected messages. So we make sure that any errors are correctly bubbled up.

**Review by nisdas — COMMENTED** [review]
```go
	// We should not attempt to process this message if the node is running in optimistic mode.
	// We just ignore in p2p so that the peer is not penalized.
	status, err := s.cfg.chain.IsOptimistic(ctx)
	if err != nil {
		return pubsub.ValidationReject, nil
	}
	if status {
		return pubsub.ValidationIgnore, nil
	}
```

This is basically used in all the message handlers, this can be abstracted to a separate method and then used in all our gossip handlers.

**Comment by terencechain** [comment]
hi @jmozah could you fix the diffs, please?

**Comment by jmozah** [comment]
Closing in favour of https://github.com/prysmaticlabs/prysm/pull/10300

**Pull request closed without merging**